### ISSUE TITLE
Version 5 for .NET 6 Support

### ISSUE THREAD

**Issue by ktoim** [json-api-dotnet/JsonApiDotNetCore#1109]
I can see in the README that this package offers .NET 6 support, however upon looking through nuget and the released on GitHub there is no version 5, I also checked branches just to be sure and can't see anything on there.

**Comment by ktoim** [comment]
Also just clarified on looking on AppVeyor to that version 5 does not exist.

**Comment by bart-degreed** [comment]
That's correct. The documentation website is published from the master branch, which is where we work on v5. We haven't tested against .NET 6 and EF Core 6 yet, but it's a goal we pursue for the stable v5 release. There's no v5 prerelease on NuGet yet, however, we welcome you to [try our latest build](https://github.com/json-api-dotnet/JsonApiDotNetCore#trying-out-the-latest-build) on .NET 6 and let us know if there are any issues.

What you can also do is try v4 on .NET 6. We'll consider officially supporting .NET 6 with v4 if the fixes are small enough.

**Comment by ktoim** [comment]
Hi bart, thank you for getting back to me and clarifying. 

I'm on my phone currently but when trying to run earlier; I couldn't get it to run because of .addJsonApi() so it fell at the first hurdle unfortunately. 

I'm willing to be a tester for this; so I will find the error later and see if you can advise anything if that's okay?

**Comment by ktoim** [comment]
So, obviously in .NET 6 the startup has changed.

It appears `builder.services.AddJsonApi<AppDbContext>();` is what the issue is relating too.

> Unhandled exception. System.MissingMethodException: Method not found: 'System.Type Microsoft.EntityFrameworkCore.Metadata.ITypeBase.get_ClrType()'.

**Comment by ktoim** [comment]
Further:

Just as an added, when I pulled the project down and referenced it locally to use

- EFCoreVersion: 6.0.0
- NetCoreAppVersion: net6.0

It worked and from testing so far, everything seems to be good! Would it be possible to get a version of .net6.0 on the AppVeyor feed?

Many thanks!

**Comment by bart-degreed** [comment]
For the exception you mentioned, can you include the full stack trace? And include which versions of .NET, EF Core, JADNC, and the database provider you used when you got this failure?

About the AppVeyor feed: we use the free tier, which means jobs cannot be run in parallel and each job is restricted to a maximum duration of 1 hour. Running the cibuild already takes a while today (and occasionally times out, the VMs are quite slow), so I'm not looking forward to doubling our build times. However, before crafting a new release I usually run some tests locally against the latest versions of frameworks, just to make sure the basics still work.

**Comment by ktoim** [comment]
@bart-degreed yeah sure, no problem! Thank you for your support on this. 

Is there a discord server for this package? Don't worry if not, just might be helpful.

So versions are as follows:

.NET: net6.0
EF Core: 6.0.0

In terms of version for the JADNC, it was the latest master branch so I believe that to be `4.2.0-pre-0855`. Hope this helps, I've been actively developing on it since posting that 2 hours ago and not encountered anything thus far.

**Comment by bart-degreed** [comment]
We use [gitter chat](https://gitter.im/json-api-dotnet-core/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) instead of discord.

Thanks for providing the details. I did some experiments myself. First, I created a simple .NET 6 WebApi project by following the steps from [Getting Started](https://www.jsonapi.net/getting-started/step-by-step.html) and ended up with the following in `Program.cs`:

```c#
using DemoWebApiOnDotNet6.Data;
using DemoWebApiOnDotNet6.Resources;
using JsonApiDotNetCore.Configuration;
using Microsoft.EntityFrameworkCore;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddDbContext<AppDbContext>(options => options.UseNpgsql(
    "Host=localhost;Port=5432;Database=DotNet6Example;User ID=postgres;Password=postgres"));

builder.Services.AddJsonApi<AppDbContext>();

var app = builder.Build();

app.UseRouting();
app.UseJsonApi();
app.MapControllers();

using (var scope = app.Services.CreateScope())
{
    var dbContext = scope.ServiceProvider.GetRequiredService<AppDbContext>();
    dbContext.Database.EnsureDeleted();
    dbContext.Database.EnsureCreated();

    dbContext.People.Add(new Person { Name = "John Doe" });
    dbContext.SaveChanges();
}

app.Run();
```

For this demo API, I found the following:
- NuGet reference to JADNC v4.2 on .NET 6 and EF Core 3.1 using PostgreSQL: works
- NuGet reference to JADNC v4.2 on .NET 6 and EF Core 5 using PostgreSQL: works
- NuGet reference to JADNC v4.2 on .NET 6 and EF Core 6 using PostgreSQL: runtime exception at application startup, because EF Core 6 has binary breaking changes

Next, I opened the solution from the [master branch](https://github.com/json-api-dotnet/JsonApiDotNetCore/commit/b7930dee25fd506fd4beb55f5b3f4280f08f3244) and found the following:
- Switched to .NET 6 with EF Core 5 using PostgreSQL: all tests pass, despite nullability warnings during build, due to richer annotations and fixes in ASP.NET libraries
- Switched to .NET 6 with EF Core 6 using PostgreSQL: 413/1729 tests fail, more nullability warnings because EF Core 6 got annotated too. After adding global `AppContext.SetSwitch("Npgsql.EnableLegacyTimestampBehavior", true);` as suggested [here](https://www.npgsql.org/efcore/release-notes/6.0.html), still 66 failing tests.

Then I opened the solution from the [4.2 release branch](https://github.com/json-api-dotnet/JsonApiDotNetCore/tree/release/4.2.0) and found the following:
- Switched to .NET 6 with EF Core 3.1 using PostgreSQL: all tests pass
- Switched to .NET 6 with EF Core 5 using PostgreSQL: all tests pass
- Switched to .NET 6 with EF Core 6: 367/1859 tests fail, after setting `EnableLegacyTimestampBehavior` all tests succeed

Conclusion: Using .NET 6 seems to work fine, EF Core 6 has issues that we'll need to investigate. Note that all the above applies to a fresh start with an empty database, which is usually not very realistic. The release notes of PostgreSQL sum it up pretty well:

> Npgsql 6.0 brings some major breaking changes and is not a simple in-place upgrade. Carefully read the breaking change notes below and upgrade with care.

**Comment by ktoim** [comment]
@bart-degreed i forgot to mention I was actually using SQLServer for my current application and didn't encounter errors. 

What do you recommend moving forward? Keep the latest version up to date locally so it's eligible for use? 

I haven't encountered any issues thus far but am keen to explore and upgrade my application to .NET 6 😀.

**Comment by bart-degreed** [comment]
Thanks for trying out the latest bits. We have a rule that each PR merge should improve the codebase, so we won't be merging half work to master. This means you can safely use the master branch until we have a prerelease out. However, you may face breaking changes until the stable v5, which we generally document in the PR descriptions. Aside from the general [list of breaking changes in EF Core 6](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-6.0/breaking-changes), I haven't found documents online for the SQL Server provider, so you should be fine.

I've been thinking about what we can do for JADNC v4. It seems fine to support .NET 6, but not EF Core 6 due to its binary breaking changes. Even if we multi-targeted JADNC against .NET 5 with EF Core 5 and .NET 6 with EF Core 6, it would mean that using .NET 6 with EF Core 5 is no longer possible because NuGet does not support forcing a different TFM  (see [here](https://stackoverflow.com/questions/49598467/how-to-choose-target-framework-from-nuget-package) and [here](https://github.com/dotnet/sdk/issues/1791)). So this would be a breaking change in JADNC v4 for existing users that have already upgraded to .NET 6, which we cannot do.

In the master branch, we currently target .NET 5 with EF Core 5, but we aim to support running against .NET 6 and EF Core 6 in the first prerelease. We could achieve supporting both the 5.x versions and the 6.x versions using multi-targeting, but given the various breaking changes in JADNC v5 already, I think it's reasonable to support just .NET 6 with EF Core 6. Yet we may still choose to multi-target, based on substantial community feedback. However, we'd like to wait a few weeks before updating master, until the first maintenance fixes for .NET 6 and EF Core 6 are out, as well as tooling updates (Resharper is currently in EAP) and dependent packages. While it's tempting to develop against the latest and greatest, dive into all sorts of bugs, create issues, workarounds, etc, we're expected to remain productive. Though I'll keep an eye out for what's coming, as well as listen to the JADNC community feedback.

I hope this gives some perspective on what to expect in the near future, as well as invite users to provide feedback on our plans.

**Comment by ktoim** [comment]
@bart-degreed that's amazing thank you for getting back in touch with me and taking the time to investigate.

Is there a place where I can **if** I experience issues with JADNC to post. Wherever is best for yourself for visibility.

Much appreciated.

**Comment by bart-degreed** [comment]
Opening separate issues for different problems would work best. It enables us to solve them separately and link PRs to them.

**Comment by ktoim** [comment]
@bart-degreed Not a problem! I'll go ahead and close this issue and raise anymore that I may come across.

Thank you for your brilliant work on the package!

**Comment by bart-degreed** [comment]
Reopening to track the work needed for .NET 6 support. Linked from roadmap to here.

**Comment by ThomasBarnekow** [comment]
Hi @bart-degreed,

> Running the cibuild already takes a while today (and occasionally times out, the VMs are quite slow)

That is definitely true. One set of activities that takes relatively long is this:

```
RunInspectCode
RunCleanupCode
```

On Linux (Ubuntu image), `RunInspectCode` takes a whopping **16 to 18 minutes** (yes, minutes) whereas it "only" takes 4 to 5 minutes on Windows (Visual Studio 2019 image). `RunCleanupCode` takes 2 to 4 minutes on Linux and 1 to 2 minutes on Windows. In total, we are talking about roughly 18 to 22 minutes on Linux and 5 to 7 minutes on Windows. Overall `RunInspectCode` accounts for roughly half of the total time on Linux.

The question is why you need to run:

- **both steps** as part of the cibuild and
- those two steps in **both images** (Ubuntu and Visual Studio 2019).

If possible, you should only run this on Windows.

**Comment by bart-degreed** [comment]
We need to run Linux before Windows, because the Windows build publishes the NuGet package and the documentation website. We'd like to fail fast on formatting issues, not wait for the whole Linux to completely build first.

**Comment by ThomasBarnekow** [comment]
I'll open a separate issue for discussing the build topic because you are right: it has been the most frustrating experience.

**Comment by ziali088** [comment]
> Unhandled exception. System.MissingMethodException: Method not found: 'System.Type Microsoft.EntityFrameworkCore.Metadata.ITypeBase.get_ClrType()'.

I've also come across this error when starting a new .NET 6 project, downgrading to EF 5 worked.

**Comment by bart-degreed** [comment]
@ziali088 See https://github.com/json-api-dotnet/JsonApiDotNetCore#compatibility. You'll need to clone/download the repo and build from source to use EF Core 6 currently.
